bigml/lda.py

```diff
@@ -0,0 +1,310 @@
+        """Returns the distribution of topics given the input text.
+
+        """
+        # Checks and cleans input_data leaving the fields used in the model
+        input_data = self.filter_input_data(input_data, by_name=by_name)
+
+        # Checks that all numeric fields are present in input data
+        for field_id, field in self.fields.items():
+            if field_id not in input_data:
+                raise Exception("Failed to predict a topic distribution.  "
+                                "Input data must contain values for all "
+                                "modeled text fields.")
+
+        return self._distribution_for_text("\n\n".join(input_data.values()))
+
+    def _distribution_for_text(self, text):
+        if isinstance(text, (str, unicode)):
+            astr = text
+        else:
+            # List of strings
+            astr = "\n\n".join(text)
+
+        doc = self._tokenize(astr)
+        return self._infer(doc, UPDATES)
+
+    def _stem(self, term):
+        if not self._stemmer:
+            return turn
+        else:
+            return self._stemmer.stemWord(term)
+
+    def _tokenize(self, astr):
+        out_terms = []
+
+        last_term = None
+        term_before = None
+
+        space_was_sep = False
+        saw_char = False
+
+        ustr = unicode(astr)
+        i = 0
+
+        while i < len(ustr):
+            if (self._bigrams and
+                last_term is not None and
+                term_before is not None):
+
+                bigram = self._stem(term_before + " " + last_term)
+                if bigram in self._term_to_idx:
+                    out_terms.append(self._term_to_idx[bigram])
+
+            char = ustr[i]
+            buf = array.array('u')
+            saw_char = False
+
+            if not char.isalnum():
+                saw_char = True
+
+            while not char.isalnum() and i < len(ustr):
+                i += 1
+                char = ustr[i]
+
+            while (i < len(ustr) and
+                   (char.isalnum() or char == "'") and
+                   len(buf) < MAX_TERM_LENGTH):
+
+                buf.append(char)
+                i += 1
+
+                if i < len(ustr):
+                    char = ustr[i]
+                else:
+                    char = None
+
+            if len(buf) > 0:
+                term_out = buf.tounicode()
+
+                if not self._case_sensitive:
+                    term_out = term_out.lower()
+
+                if space_was_sep and not saw_char:
+                    term_before = last_term
+                else:
+                    term_before = None
+
+                last_term = term_out
+
+                if char == " " or char == "\n":
+                    space_was_sep = True
+
+                tstem = self._stem(term_out)
+                if tstem in self._term_to_idx:
+                    out_terms.append(self._term_to_idx[tstem])
+
+                i += 1
+
+        if (self._bigrams and
```

Comment:
If I was jao, I'd ask you to build a function to do this segment of code, whiich is repeated in the while loop and here...